File: packages/php/src/Value/DocumentMetadata.php
<?php

declare(strict_types=1);

namespace HtmlToMarkdown\Value;

use HtmlToMarkdown\Internal\TypeAssertions;

final readonly class DocumentMetadata
{
    /**
     * @param array<string, string> $openGraph
     * @param array<string, string> $twitterCard
     * @param array<string, string> $metaTags
     * @param list<string> $keywords
     */
    public function __construct(
        public ?string $title,
        public ?string $description,
        public array $keywords,
        public ?string $author,
        public ?string $canonicalUrl,
        public ?string $baseHref,
        public ?string $language,
        public ?string $textDirection,
        public array $openGraph,
        public array $twitterCard,
        public array $metaTags,
    ) {
    }

    /**
     * @param array<string, mixed> $payload
     */
    public static function fromExtensionPayload(array $payload): self
    {
        return new self(
            title: TypeAssertions::stringOrNull($payload['title'] ?? null, 'document_metadata.title'),
            description: TypeAssertions::stringOrNull($payload['description'] ?? null, 'document_metadata.description'),
            keywords: \array_key_exists('keywords', $payload)
                ? TypeAssertions::stringList($payload['keywords'], 'document_metadata.keywords')
                : [],
            author: TypeAssertions::stringOrNull($payload['author'] ?? null, 'document_metadata.author'),
            canonicalUrl: TypeAssertions::stringOrNull(
                $payload['canonical_url'] ?? null,
                'document_metadata.canonical_url',
            ),
            baseHref: TypeAssertions::stringOrNull($payload['base_href'] ?? null, 'document_metadata.base_href'),
            language: TypeAssertions::stringOrNull($payload['language'] ?? null, 'document_metadata.language'),
            textDirection: TypeAssertions::stringOrNull(
                $payload['text_direction'] ?? null,
                'document_metadata.text_direction',
            ),
            openGraph: TypeAssertions::stringMap($payload['open_graph'] ?? [], 'document_metadata.open_graph'),
            twitterCard: TypeAssertions::stringMap($payload['twitter_card'] ?? [], 'document_metadata.twitter_card'),
            metaTags: TypeAssertions::stringMap($payload['meta_tags'] ?? [], 'document_metadata.meta_tags'),
        );
    }
}

File: packages/java/src/main/java/dev/kreuzberg/htmltomarkdown/metadata/DocumentMetadata.java
package dev.kreuzberg.htmltomarkdown.metadata;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Map;

/**
 * Document-level metadata extracted from {@code <head>} and top-level elements.
 *
 * <p>Contains all metadata typically used by search engines, social media platforms, and browsers
 * for document indexing and presentation.
 *
 * @param title Document title from {@code <title>} tag
 * @param description Document description from {@code <meta name="description">} tag
 * @param keywords Document keywords from {@code <meta name="keywords">} tag, split on commas
 * @param author Document author from {@code <meta name="author">} tag
 * @param canonicalUrl Canonical URL from {@code <link rel="canonical">} tag
 * @param baseHref Base URL from {@code <base href="">} tag for resolving relative URLs
 * @param language Document language from {@code lang} attribute
 * @param textDirection Document text direction from {@code dir} attribute
 * @param openGraph Open Graph metadata (og:* properties) for social media
 * @param twitterCard Twitter Card metadata (twitter:* properties)
 * @param metaTags Additional meta tags not covered by specific fields
 * @since 2.13.0
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record DocumentMetadata(
    @JsonProperty("title") String title,
    @JsonProperty("description") String description,
    @JsonProperty("keywords") java.util.List<String> keywords,
    @JsonProperty("author") String author,
    @JsonProperty("canonical_url") String canonicalUrl,
    @JsonProperty("base_href") String baseHref,
    @JsonProperty("language") String language,
    @JsonProperty("text_direction") TextDirection textDirection,
    @JsonProperty("open_graph") Map<String, String> openGraph,
    @JsonProperty("twitter_card") Map<String, String> twitterCard,
    @JsonProperty("meta_tags") Map<String, String> metaTags) {

  /**
   * Construct a DocumentMetadata record with all fields.
   *
   * @param title the document title
   * @param description the document description
   * @param keywords the list of keywords
   * @param author the document author
   * @param canonicalUrl the canonical URL
   * @param baseHref the base href
   * @param language the document language
   * @param textDirection the text direction
   * @param openGraph the open graph properties
   * @param twitterCard the twitter card properties
   * @param metaTags additional meta tags
   */
  public DocumentMetadata {
    keywords = keywords != null ? keywords : java.util.Collections.emptyList();
    openGraph = openGraph != null ? openGraph : java.util.Collections.emptyMap();
    twitterCard = twitterCard != null ? twitterCard : java.util.Collections.emptyMap();
    metaTags = metaTags != null ? metaTags : java.util.Collections.emptyMap();
  }

  /**
   * Create a DocumentMetadata with minimal information.
   *
   * @param title the document title
   * @param description the document description
   * @return a new DocumentMetadata instance
   */
  public static DocumentMetadata of(String title, String description) {
    return new DocumentMetadata(
        title,
        description,
        java.util.Collections.emptyList(),
        null,
        null,
        null,
        null,
        null,
        java.util.Collections.emptyMap(),
        java.util.Collections.emptyMap(),
        java.util.Collections.emptyMap());
  }
}
